cmd/sops/subcommand/publish/publish.go

```diff
@@ -0,0 +1,114 @@
+package publish
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"go.mozilla.org/sops"
+	"go.mozilla.org/sops/cmd/sops/codes"
+	"go.mozilla.org/sops/cmd/sops/common"
+	"go.mozilla.org/sops/config"
+	"go.mozilla.org/sops/keyservice"
+	"go.mozilla.org/sops/logging"
+
+	"github.com/sirupsen/logrus"
+)
+
+var log *logrus.Logger
+
+func init() {
+	log = logging.NewLogger("PUBLISH")
+}
+
+type Opts struct {
+	Interactive bool
+	ConfigPath  string
+	InputPath   string
+	KeyServices []keyservice.KeyServiceClient
+	InputStore  sops.Store
+}
+
+func Run(opts Opts) error {
+	var fileContents []byte
+	path, err := filepath.Abs(opts.InputPath)
+	if err != nil {
+		return err
+	}
+	info, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		return fmt.Errorf("can't operate on a directory")
+	}
+	_, fileName := filepath.Split(path)
+
+	conf, err := config.LoadForFile(opts.ConfigPath, opts.InputPath, make(map[string]*string))
+	if err != nil {
+		return err
+	}
+	if conf.Destination == nil {
+		return errors.New("No destination configured for this file.")
+	}
+
+	// Check that this is a sops-encrypted file
+	tree, err := common.LoadEncryptedFile(opts.InputStore, opts.InputPath)
+	if err != nil {
+		return err
+	}
+
+	// Re-encrypt if settings exist to do so
+	if len(conf.ReEncryptionKeyGroups) != 0 {
+		key, err := tree.Metadata.GetDataKeyWithKeyServices(opts.KeyServices)
+		if err != nil {
+			return common.NewExitError(err, codes.CouldNotRetrieveKey)
+		}
+		tree.Metadata.KeyGroups = conf.ReEncryptionKeyGroups
+		tree.Metadata.ShamirThreshold = min(tree.Metadata.ShamirThreshold, len(tree.Metadata.KeyGroups))
```

Comment:
I don't think this should silently pick the lowest value between number of keygroups and shamir threshold. If a shamir threshold is configured with a value > 1, and there are not enough keygroups to satisfy the requirement, then we should exit with an error.

Reply:
I think it might be best to let users define what Shamir threshold they want in their reencryption rules. Essentially, I'd make the reencryption rules have the same format as normal creation rules, which already let you specify a Shamir threshold.